tools/rosclean/src/rosclean/__init__.py

```diff
@@ -138,20 +152,68 @@ def get_disk_usage(d):
     else:
         raise CleanupException("rosclean is not supported on this platform")
 
+def _sort_file_by_oldest(d):
+    """
+    Get files and directories in specified path sorted by last modified time
+    :param d: directory path, ```str```
+    :return:  a list of files and directories sorted by last modified time (old first), ```list```
+    """
+    files = os.listdir(d)
+    files.sort(key=lambda file: os.path.getmtime(os.path.join(d,file)))
+    return files
+
+def _check_delete_file_size(d, file):
+    """
+    Get size of specified directory after deleting specified file or directory
+    :param d: directory path, ```str```
+    :param file: file or directory, ```str```
+    :return: size of directory after purging file, ```int```
+    """
+    d_size = get_disk_usage(d)
+    target = os.path.join(d, file)
+    if os.path.isdir(target):
+        files = os.listdir(target)
+        f_size = 0
+        for f in files:
+            f_size = f_size + os.path.getsize(os.path.join(target, f))
+    else:
+        f_size = os.path.getsize(target)
+    return (d_size - f_size)
+
 def _rosclean_cmd_purge(args):
     dirs = _get_check_dirs()
 
     for d, label in dirs:
-        print("Purging %s."%label)
-        cmds = [['rm', '-rf', d]]
-        try:
-            if args.y:
-                _call(cmds)
-            else:
+        if not args.size:
+            print("Purging %s."%label)
+            cmds = [['rm', '-rf', d]]
+            try:
+                if args.y:
+                    _call(cmds)
+                else:
+                    print("PLEASE BE CAREFUL TO VERIFY THE COMMAND BELOW!")
+                    _ask_and_call(cmds)
+            except:
+                print("FAILED to execute command", file=sys.stderr)
+        else:
+            print("Purging %s until directory size is %d MB."%(label, args.size))
+            if not args.y:
                 print("PLEASE BE CAREFUL TO VERIFY THE COMMAND BELOW!")
-                _ask_and_call(cmds)
-        except:
-            print("FAILED to execute command", file=sys.stderr)
+                exe = _ask("Purge some of old logs in %s"%d)
+            else:
+                exe = True
+            if exe:
+                files = _sort_file_by_oldest(d)
+                for file in files:
+                    if _check_delete_file_size(d, file) >= (args.size * 1024 * 1024):
```

Comment:
Currently the algorithm calls `get_disk_usage()` repeatedly. I would suggest to only compute it once. And then for item being considered to be removed only calculate the size of that item. The new remaining size can then be computed from these two in each round of the loop.